### PR TITLE
Add from_origin to trunk_vectors and related features

### PR DESCRIPTION
**Pull request by arnaudon**
In some usecases, for example in synthesis, the reference point to compute trunk angles should be [0, 0, 0] instead of soma.center.

See https://github.com/BlueBrain/NeuroTS/pull/49/files#diff-a465052b0c2a633fd29c380c8f81a1143bc78e09982d08b847901e453ab99f4eR105

### CODE REVIEW

File: neurom/features/morphology.py [hunk 1]
```diff
@@ -64,9 +64,7 @@
 
 
 def _map_neurites(function, morph, neurite_type):
-    return list(
-        iter_neurites(morph, mapfun=function, filt=is_type(neurite_type))
-    )
+    return list(iter_neurites(morph, mapfun=function, filt=is_type(neurite_type)))
```
Comment (arnaudon) [inline neurom/features/morphology.py:67]:
```suggestion
    return list(
        iter_neurites(morph, mapfun=function, filt=is_type(neurite_type))
    )
```

File: neurom/features/morphology.py [hunk 2]
```diff
@@ -82,7 +80,7 @@ def soma_surface_area(morph):
     Note:
         The surface area is calculated by assuming the soma is spherical.
     """
-    return 4.0 * math.pi * morph.soma.radius ** 2
+    return 4.0 * math.pi * morph.soma.radius**2
```
Comment (arnaudon) [inline neurom/features/morphology.py:83]:
```suggestion
    return 4.0 * math.pi * morph.soma.radius ** 2
```

File: neurom/features/morphology.py [hunk 3]
```diff
@@ -132,6 +130,7 @@ def trunk_origin_azimuths(morph, neurite_type=NeuriteType.all):
 
     The range of the azimuth angle [-pi, pi] radians
     """
+
```
Comment (arnaudon) [inline neurom/features/morphology.py:133]:
```suggestion
```

File: neurom/features/morphology.py [hunk 4]
```diff
@@ -151,6 +150,7 @@ def trunk_origin_elevations(morph, neurite_type=NeuriteType.all):
 
     The range of the elevation angle [-pi/2, pi/2] radians
     """
+
```
Comment (arnaudon) [inline neurom/features/morphology.py:153]:
```suggestion
```

File: neurom/features/morphology.py [hunk 5]
```diff
@@ -213,7 +224,8 @@ def trunk_angles(
                     morphmath.angle_between_projections(i / np.linalg.norm(i), [0, 1])
                     for i in vectors[:, sort_coords]
                 ),
-                dtype=float)
+                dtype=float,
+            )
```
Comment (arnaudon) [inline neurom/features/morphology.py:228]:
```suggestion
                dtype=float)
```

File: neurom/features/morphology.py [hunk 6]
```diff
@@ -226,10 +238,13 @@ def trunk_angles(
...
-            morphmath.angle_between_vectors(i, j)
-            for j in cycling_vectors[num_i: num_i + n_vectors]
-        ])
+        (
+            num_i,
+            [
+                morphmath.angle_between_vectors(i, j)
+                for j in cycling_vectors[num_i : num_i + n_vectors]
+            ],
+        )
```
Comment (arnaudon) [inline neurom/features/morphology.py:247]:
```suggestion
        (num_i, [
            morphmath.angle_between_vectors(i, j)
            for j in cycling_vectors[num_i: num_i + n_vectors]
        ])
```

File: neurom/features/morphology.py [hunk 7]
```diff
@@ -286,9 +303,9 @@ def trunk_angles_inter_types(
     for i, source in enumerate(source_vectors):
         for j, target in enumerate(target_vectors):
             angles[i, j, 0] = morphmath.angle_between_vectors(source, target)
-            angles[i, j, [1, 2]] = (
-                morphmath.spherical_from_vector(target) - morphmath.spherical_from_vector(source)
-            )
+            angles[i, j, [1, 2]] = morphmath.spherical_from_vector(
+                target
+            ) - morphmath.spherical_from_vector(source)
```
Comment (arnaudon) [inline neurom/features/morphology.py:308]:
```suggestion
            angles[i, j, [1, 2]] = (
                morphmath.spherical_from_vector(target) - morphmath.spherical_from_vector(source)
            )
```

File: neurom/features/morphology.py [hunk 8]
```diff
@@ -298,8 +315,7 @@ def trunk_angles_inter_types(
 
     if closest_component is not None:
         angles = angles[
-            np.arange(len(angles)),
-            np.argmin(np.abs(angles[:, :, closest_component]), axis=1)
+            np.arange(len(angles)), np.argmin(np.abs(angles[:, :, closest_component]), axis=1)
```
Comment (arnaudon) [inline neurom/features/morphology.py:318]:
```suggestion
            np.arange(len(angles)),
            np.argmin(np.abs(angles[:, :, closest_component]), axis=1)
```

**Comment by eleftherioszisis** [comment]
Can you please revert the formatting? The code has already been formatted on the v4 branch and this will only cause conflicts. :)

File: neurom/features/morphology.py [hunk 9]
```diff
@@ -338,9 +356,9 @@ def trunk_angles_from_vector(
     angles = np.empty((len(vectors), 3), dtype=float)
     for i, i_vec in enumerate(vectors):
         angles[i, 0] = morphmath.angle_between_vectors(vector, i_vec)
-        angles[i, (1, 2)] = (
-            morphmath.spherical_from_vector(i_vec) - morphmath.spherical_from_vector(vector)
-        )
+        angles[i, (1, 2)] = morphmath.spherical_from_vector(
+            i_vec
+        ) - morphmath.spherical_from_vector(vector)
```
Comment (arnaudon) [inline neurom/features/morphology.py:361]:
```suggestion
        angles[i, (1, 2)] = (
            morphmath.spherical_from_vector(i_vec) - morphmath.spherical_from_vector(vector)
        )
```

File: neurom/features/morphology.py [hunk 10]
```diff
@@ -380,8 +398,9 @@ def trunk_origin_radii(
               ``max_length_filter`` are returned.
     """
     if max_length_filter is None and min_length_filter is None:
-        return [n.root_node.points[0][COLS.R]
-                for n in iter_neurites(morph, filt=is_type(neurite_type))]
+        return [
+            n.root_node.points[0][COLS.R] for n in iter_neurites(morph, filt=is_type(neurite_type))
+        ]
```
Comment (arnaudon) [inline neurom/features/morphology.py:403]:
```suggestion
        return [n.root_node.points[0][COLS.R]
                for n in iter_neurites(morph, filt=is_type(neurite_type))]
```

File: neurom/features/morphology.py [hunk 11]
```diff
@@ -411,7 +430,7 @@ def _mean_radius(neurite):
         path_lengths = np.insert(np.cumsum(interval_lengths), 0, 0)
         valid_pts = np.ones(len(path_lengths), dtype=bool)
         if min_length_filter is not None:
-            valid_pts = (valid_pts & (path_lengths >= min_length_filter))
+            valid_pts = valid_pts & (path_lengths >= min_length_filter)
```
Comment (arnaudon) [inline neurom/features/morphology.py:433]:
```suggestion
            valid_pts = (valid_pts & (path_lengths >= min_length_filter))
```

File: neurom/features/morphology.py [hunk 12]
```diff
@@ -420,8 +439,8 @@ def _mean_radius(neurite):
                 )
                 return points[-1, COLS.R]
         if max_length_filter is not None:
-            valid_max = (path_lengths <= max_length_filter)
-            valid_pts = (valid_pts & valid_max)
+            valid_max = path_lengths <= max_length_filter
+            valid_pts = valid_pts & valid_max
```
Comment (arnaudon) [inline neurom/features/morphology.py:443]:
```suggestion
            valid_max = (path_lengths <= max_length_filter)
            valid_pts = (valid_pts & valid_max)
```

File: neurom/features/morphology.py [hunk 13]
```diff
@@ -438,6 +457,7 @@ def _mean_radius(neurite):
 @feature(shape=(...,))
 def trunk_section_lengths(morph, neurite_type=NeuriteType.all):
     """List of lengths of trunk sections of neurites in a morph."""
+
```
Comment (arnaudon) [inline neurom/features/morphology.py:460]:
```suggestion
```

File: neurom/features/morphology.py [hunk 14]
```diff
@@ -478,30 +498,37 @@ def sholl_crossings(morph, neurite_type=NeuriteType.all, center=None, radii=None
                                                         center=morph.soma.center,
                                                         radii=np.arange(0, 1000, 100))
     """
+
```
Comment (arnaudon) [inline neurom/features/morphology.py:501]:
```suggestion
```

File: neurom/features/morphology.py [hunk 15]
```diff
@@ -478,30 +498,37 @@ def sholl_crossings(morph, neurite_type=NeuriteType.all, center=None, radii=None
...
     def _count_crossings(neurite, radius):
         """Used to count_crossings of segments in neurite with radius."""
-        r2 = radius ** 2
+        r2 = radius**2
```
Comment (arnaudon) [inline neurom/features/morphology.py:504]:
```suggestion
        r2 = radius ** 2
```

File: neurom/features/morphology.py [hunk 16]
```diff
@@ -478,30 +498,37 @@ def sholl_crossings(morph, neurite_type=NeuriteType.all, center=None, radii=None
...
         count = 0
         for start, end in iter_segments(neurite):
-            start_dist2, end_dist2 = (morphmath.point_dist2(center, start),
-                                      morphmath.point_dist2(center, end))
+            start_dist2, end_dist2 = (
+                morphmath.point_dist2(center, start),
+                morphmath.point_dist2(center, end),
+            )
```
Comment (arnaudon) [inline neurom/features/morphology.py:510]:
```suggestion
            start_dist2, end_dist2 = (morphmath.point_dist2(center, start),
                                      morphmath.point_dist2(center, end))
```

File: neurom/features/morphology.py [hunk 17]
```diff
@@ -478,30 +498,37 @@ def sholl_crossings(morph, neurite_type=NeuriteType.all, center=None, radii=None
...
 
-            count += int(start_dist2 <= r2 <= end_dist2 or
-                         end_dist2 <= r2 <= start_dist2)
+            count += int(start_dist2 <= r2 <= end_dist2 or end_dist2 <= r2 <= start_dist2)
```
Comment (arnaudon) [inline neurom/features/morphology.py:512]:
```suggestion
            count += int(start_dist2 <= r2 <= end_dist2 or
                         end_dist2 <= r2 <= start_dist2)
```

File: neurom/features/morphology.py [hunk 18]
```diff
@@ -478,30 +498,37 @@ def sholl_crossings(morph, neurite_type=NeuriteType.all, center=None, radii=None
...
         return count
 
     if center is None or radii is None:
-        assert isinstance(morph, Morphology) and morph.soma, \
-            '`sholl_crossings` input error. If `center` or `radii` is not set then `morph` is ' \
-            'expected to be an instance of Morphology and have a soma.'
+        assert isinstance(morph, Morphology) and morph.soma, (
+            "`sholl_crossings` input error. If `center` or `radii` is not set then `morph` is "
+            "expected to be an instance of Morphology and have a soma."
+        )
```
Comment (arnaudon) [inline neurom/features/morphology.py:520]:
```suggestion
        assert isinstance(morph, Morphology) and morph.soma, \
            '`sholl_crossings` input error. If `center` or `radii` is not set then `morph` is ' \
            'expected to be an instance of Morphology and have a soma.'
```

File: neurom/features/morphology.py [hunk 19]
```diff
@@ -478,30 +498,37 @@ def sholl_crossings(morph, neurite_type=NeuriteType.all, center=None, radii=None
...
-    return [sum(_count_crossings(neurite, r)
-                for neurite in iter_neurites(morph, filt=is_type(neurite_type)))
-            for r in radii]
+    return [
+        sum(
+            _count_crossings(neurite, r)
+            for neurite in iter_neurites(morph, filt=is_type(neurite_type))
+        )
+        for r in radii
+    ]
```
Comment (arnaudon) [inline neurom/features/morphology.py:531]:
```suggestion
    return [sum(_count_crossings(neurite, r)
                for neurite in iter_neurites(morph, filt=is_type(neurite_type)))
            for r in radii]
```

File: neurom/features/morphology.py [hunk 20]
```diff
@@ -532,7 +559,8 @@ def sholl_frequency(morph, neurite_type=NeuriteType.all, step_size=10, bins=None
 
         max_radius_per_neurite = [
             np.max(np.linalg.norm(n.points[:, COLS.XYZ] - morph.soma.center, axis=1))
-            for n in morph.neurites if neurite_filter(n)
+            for n in morph.neurites
+            if neurite_filter(n)
```
Comment (arnaudon) [inline neurom/features/morphology.py:563]:
```suggestion
            for n in morph.neurites if neurite_filter(n)
```

File: neurom/features/morphology.py [hunk 21]
```diff
@@ -550,9 +578,7 @@ def _extent_along_axis(morph, axis, neurite_type):
     along the coordinate axis direction (e.g. COLS.X).
     """
     it_points = (
-            p
-            for n in iter_neurites(morph, filt=is_type(neurite_type))
-            for p in n.points[:, axis]
+        p for n in iter_neurites(morph, filt=is_type(neurite_type)) for p in n.points[:, axis]
```
Comment (arnaudon) [inline neurom/features/morphology.py:581]:
```suggestion
            p
            for n in iter_neurites(morph, filt=is_type(neurite_type))
            for p in n.points[:, axis]
```

File: neurom/features/morphology.py [hunk 22]
```diff
@@ -609,7 +635,7 @@ def volume_density(morph, neurite_type=NeuriteType.all):
     return total_volume / morph_hull.volume
 
 
-def _unique_projected_points(morph, projection_plane,  neurite_type):
+def _unique_projected_points(morph, projection_plane, neurite_type):
```
Comment (arnaudon) [inline neurom/features/morphology.py:638]:
```suggestion
def _unique_projected_points(morph, projection_plane,  neurite_type):
```

File: neurom/features/morphology.py [hunk 23]
```diff
@@ -623,9 +649,7 @@ def _unique_projected_points(morph, projection_plane,  neurite_type):
             f"Please select 'xy', 'xz', or 'yz'."
         ) from e
 
-    points = list(
-        iter_neurites(morph, mapfun=sf.section_points, filt=is_type(neurite_type))
-    )
+    points = list(iter_neurites(morph, mapfun=sf.section_points, filt=is_type(neurite_type)))
```
Comment (arnaudon) [inline neurom/features/morphology.py:652]:
```suggestion
    points = list(
        iter_neurites(morph, mapfun=sf.section_points, filt=is_type(neurite_type))
    )
```

**Comment by eleftherioszisis** [comment]
> In some usecases, for example in synthesis, the reference point to compute trunk angles should be [0, 0, 0] instead of soma.center.

Hmm, this doesn't make much sense, does it?

Trunk angles are the angles of a trunk on the soma. It just happens in synthesis for the soma to be seated on the origin, i.e. [0,0,0]

**Comment by arnaudon** [comment]
> > In some usecases, for example in synthesis, the reference point to compute trunk angles should be [0, 0, 0] instead of soma.center.
> 
> Hmm, this doesn't make much sense, does it?
> 
> Trunk angles are the angles of a trunk on the soma. It just happens in synthesis for the soma to be seated on the origin, i.e. [0,0,0]

The trick is that in synthesis, we place trunks wrt [0, 0, 0], but the resulting soma is not spherical/symmetric, and soma.center is thus not at [0, 0, 0], so if we measure the trunk angle post-synthesis, we don't get the angles we provided to synthesis. I'm fine to leave this inside neurots (as in the link above). We discussed with @adrien-berchet that it may be good to have this less 'hidden' in neurots here. If you don't want, it's fine I suppose. 

@lidakanari , sorry I forgot to ping you here for the discussion.

**Comment by lidakanari** [comment]
I don't think that it's a reasonable change. I think the easiest fix for this issue is related to NeuroTS and should be implemented there. As an idea: 
* given the initial center of the soma (x0,y0,z0)
* generate the soma points
* recompute new center based on points
* move the soma points at the opposite direction so that the center is again (x0,y0,z0)

**Comment by adrien-berchet** [comment]
We also thought about method but there is an issue with it: if we move all the soma points then the angles will be changed, especially the angle of the apical will not be vertical anymore. But maybe it's no big deal? Maybe the only important feature is that the first segment of the apical is vertical but we don't care about the angle between the soma center and the first point of the apical (well, it should still be quite close to vertical ofc).

**Comment by arnaudon** [comment]
This PR has never existed

**Comment by eleftherioszisis** [comment]
> This PR has never existed

What PR? :D

**Pull request closed without merging**